### PR TITLE
[batch] sql temporary tables

### PR DESCRIPTION
**Pull request by jigold**
Depends on #5776.

### CODE REVIEW

File: batch/test-locally.sh [hunk 1]
```diff
@@ -24,6 +27,8 @@ else
     export CLOUD_SQL_CONFIG_PATH=/batch-secrets/batch-test-cloud-sql-config.json
 fi
 
+temp_table=$(python3 -c "from batch.server.database import Database; db = Database.create_synchronous(\"$CLOUD_SQL_CONFIG_PATH\"); print(db.temp_table_name_sync(\"foo\"))")
```
Comment (jigold) [inline batch/test-locally.sh]: This is to put the infrastructure in for creating temporary tables and then cleaning them up. I could just generate a uid with `../generate_uid` and not check it's in the table. Would probably be simpler, but still need the trap command above. 

Open to suggestions on how to better do this.

File: batch/test/test_database.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        assert uid2.startswith(table_root)
+
+        schema = {'id': 'BIGINT'}
+        key = ['id']
+        self.db.create_table_sync(uid1, schema, key)
+        self.db.create_table_sync(uid2, schema, key)
+        assert self.db.has_table_sync(uid1)
+        assert self.db.has_table_sync(uid2)
+
+        self.db.drop_table_sync(uid1, uid2)
```
Comment (jigold) [inline batch/test/test_database.py]: Wanted to test that dropping multiple tables works!

File: batch/batch/server/database.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return run_synchronous(self.temp_table_name(root))
+
+    async def has_table(self, name):
+        async with self.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"SELECT * FROM INFORMATION_SCHEMA.tables " \
+                    f"WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s"
+                await cursor.execute(sql, (self.db, name))
+                result = cursor.fetchone()
+        return result.result() is not None
```
Comment (danking) [inline batch/batch/server/database.py]: I don't think you actually want this. Any paired use of this with create_table has a race condition. I think the right thing to do is to create a table with a random name catch any errors and try again until you succeed.

Comment (jigold) [reply batch/batch/server/database.py]: I don't see the race condition, but I'll rewrite the temporary table creation with a try /except. I think has_table is still a good method to have.

Comment (jigold) [reply batch/batch/server/database.py]: I think I see what you mean with the race condition, but the issue is I use these functions in the `test-locally.sh` file to get the table names that are not actually created until the server starts running. I didn't want to have to specify the schema and keys in the `test-locally` file to actually create the table. I can try having the server create the entire batch database and print those temp names. But then I have to figure out how to parse an array output from a Python command in Bash.

Comment (jigold) [reply batch/batch/server/database.py]: Ok. I fixed the race condition, but had to take out the test command from `test-locally`. I'll deal with that in a later PR.

File: batch/test/test_database.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    def test_create_drop(self):
+        schema = {'id': 'BIGINT'}
+        key = ['id']
+
+        t1_name = self.db.create_temp_table_sync("t1", schema, key)
+        t2_name = self.db.create_temp_table_sync("t2", schema, key)
+        assert t1_name.startswith("t1")
+        assert t2_name.startswith("t2")
+
+        self.db.drop_table_sync(t1_name, t2_name)
```
Comment (danking) [inline batch/test/test_database.py]: this should be in a finally block to ensure we always clean up

File: batch/test/test_database.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            t2_name = self.db.create_temp_table_sync("t2", schema, key)
+            assert t1_name.startswith("t1")
+            assert t2_name.startswith("t2")
+
+            self.db.drop_table_sync(t1_name, t2_name)
+            assert not self.db.has_table_sync(t1_name)
+            assert not self.db.has_table_sync(t2_name)
+        finally:
+            self.db.drop_table_sync(t1_name)
+            self.db.drop_table_sync(t2_name)
```
Comment (danking) [inline batch/test/test_database.py]: If you do these separately then the first needs a try and the latter needs to be in a finally block. I think it's better to use the two arg form.

**Pull request merged**